### PR TITLE
[V1][Spec Decode] Implement Eagle Proposer [1/N]

### PR DESCRIPTION
**Pull request by WoosukKwon**
Efficient implementation based upon #15346 

NOTE1: I intentionally used a dummy model instead of the real eagle models to reduce the scope of the PR.

NOTE2: Currently, I intentionally ignored all sampling parameters except the temperature when sampling the draft tokens.
The reason being is,
1. Ignoring the sampling parameter for draft tokens doesn’t change the final outputs after rejection sampling, only affecting the acceptance rate.
2. Given that the draft model is small, applying top-p, top-k, etc. could be relatively very expensive. I hypothesize that avoiding these expensive ops despite slightly degraded acceptance rate would lead to better end performance.

cc @sroy745 @LiuXiaoxuanPKU

### CODE REVIEW

File: vllm/v1/spec_decode/eagle.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            attn_metadata.max_seq_len += 1
+            attn_metadata.seq_lens += 1
+            attn_metadata.slot_mapping = compute_slot_mapping(
+                positions=positions,
+                block_table=block_table,
+                block_size=self.block_size,
+            )
+
+            # Run the model.
+            hidden_states = self.model(
```
Comment (njhill) [inline vllm/v1/spec_decode/eagle.py:115]: Need `with set_forward_context(attn_metadata, self.vllm_config):` here?

File: vllm/v1/spec_decode/eagle.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        if self.num_speculative_tokens == 1:
+            return draft_token_ids.view(-1, 1)
+
+        # Generate the remaining draft tokens.
+        draft_token_ids_list = [draft_token_ids]
+        positions = target_positions[last_token_indices]
+        hidden_states = sample_hidden_states
+        attn_metadata.num_actual_tokens = batch_size
+        attn_metadata.query_start_loc = torch.arange(batch_size,
+                                                     device=positions.device)
```
Comment (njhill) [inline vllm/v1/spec_decode/eagle.py]: preallocate?

File: vllm/v1/spec_decode/eagle.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    # NOTE(woosuk): We don't need to apply all the sampling parameters
+    # for generating the draft tokens.
+    if sampling_metadata.all_greedy:
+        # All greedy.
+        next_token_ids = logits.argmax(dim=-1)
+    else:
+        logits.div_(sampling_metadata.temperature)
+        probs = logits.softmax(dim=-1, dtype=torch.float32)
+
+        # TODO(woosuk): Consider seeds?
```
Comment (njhill) [inline vllm/v1/spec_decode/eagle.py:141]: I'm not sure seeds will work with conditional speculation and rejection sampling? :(

File: vllm/v1/spec_decode/eagle.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py]: Nice catch. Fixed!

File: vllm/v1/spec_decode/eagle.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py]: Oh I actually did the optimization in parallel. 👍 Fixed now.

File: vllm/v1/spec_decode/eagle.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        batch_size = next_token_ids.shape[0]
+        last_token_indices = cu_num_tokens[1:] - 1
+
+        input_ids = target_token_ids
+        # Shift the input ids by one token.
+        # E.g., [a1, b1, b2, c1, c2, c3] -> [b1, b2, c1, c2, c3, c3]
+        input_ids[:-1] = target_token_ids[1:]
+        # Replace the last token with the next token.
+        # E.g., [b1, b2, c1, c2, c3, c3] -> [a2, b2, b3, c2, c3, c4]
+        input_ids[last_token_indices] = next_token_ids
```
Comment (sroy745) [inline vllm/v1/spec_decode/eagle.py:52]: Will this work the case where the sequence is a decode sequence and the input is T1, T2, T3, T4 but we only accept T2?

File: vllm/v1/spec_decode/eagle.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+        seq_lens = target_positions[last_token_indices] + 1
+        max_seq_len = seq_lens.max().item()  # FIXME: Avoid synchronization.
+        slot_mapping = compute_slot_mapping(
```
Comment (sroy745) [inline vllm/v1/spec_decode/eagle.py:56]: Will compute the slot mapping using the block_tables work for the case when prefix cache is enabled?

File: vllm/v1/spec_decode/eagle.py (hunk 4 above)
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py:52]: Yes. I will drop the rejected tokens from `target_token_ids`.

File: vllm/v1/spec_decode/eagle.py (hunk 5 above)
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py:56]: I don't think any reason not. Could you please elaborate more why prefix caching can be concerned?

File: vllm/v1/spec_decode/eagle.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    # for generating the draft tokens.
+    if sampling_metadata.all_greedy:
+        # All greedy.
+        next_token_ids = logits.argmax(dim=-1)
+    else:
+        logits.div_(sampling_metadata.temperature.view(-1, 1))
+        probs = logits.softmax(dim=-1, dtype=torch.float32)
+
+        # TODO(woosuk): Consider seeds?
+        q = torch.empty_like(logits)
```
Comment (LiuXiaoxuanPKU) [inline vllm/v1/spec_decode/eagle.py:210]: At least we need to enable seed for reproducibility? i.e., the draft head proposes the same token when seeded.

File: vllm/v1/spec_decode/eagle.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+                    input_ids=input_ids,
+                    hidden_states=hidden_states,
+                    positions=positions,
+                )
+            logits = self.model.compute_logits(hidden_states, None)
+            draft_token_ids = sample_token_ids(logits, sampling_metadata)
+            draft_token_ids_list.append(draft_token_ids)
+
+        # [batch_size, num_speculative_tokens]
+        return torch.stack(draft_token_ids_list, dim=1)
```
Comment (LiuXiaoxuanPKU) [inline vllm/v1/spec_decode/eagle.py:126]: Just want to double check, currently, the eagle proposer only returns `draft_token_ids_list`, it does not return the `draft_probs`, therefore the rejection sampler will generate one hot decoding for proposed tokens, we will add that in the future PR?

File: vllm/v1/spec_decode/eagle.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+            positions += 1
+            attn_metadata.max_seq_len += 1
+            attn_metadata.seq_lens += 1
+            # Compute the slot mapping.
+            block_numbers = positions // self.block_size
+            block_ids = block_table.gather(dim=1,
+                                           index=block_numbers.view(-1, 1))
+            block_ids = block_ids.view(-1)
+            attn_metadata.slot_mapping = (block_ids * self.block_size +
+                                          positions % self.block_size)
```
Comment (LiuXiaoxuanPKU) [inline vllm/v1/spec_decode/eagle.py:112]: Here, we assume the blocks are already allocated for the draft tokens, did we allocate them anywhere?

**Review by LiuXiaoxuanPKU — COMMENTED** [review]
Overall looks good to me, just have some questions. Also, we might want some tests for this PR? Maybe move some from #15346 if possible?

File: vllm/v1/spec_decode/eagle.py (hunk 8 above)
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py:112]: Good question. This PR doesn't include it.

File: vllm/v1/spec_decode/eagle.py (hunk 7 above)
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py]: Great point. I've updated the code to return the draft probabilities in addition to the token ids.

**Comment by WoosukKwon** [comment]
@njhill @LiuXiaoxuanPKU @sroy745 Thanks for the reviews.

Because the integration of EAGLE involves a lot of changes, I tried to break it down to smaller steps. This PR is the first step to establish the high-level design, so I intentionally left many TODOs and FIXMEs.

The next steps should be:
1. Correctly initializing and loading the EAGLE draft model
2. Consider the lookahead slots in the KV cache manager
3. Cache `draft_probs` inside the model runner and correctly feed it to the rejection sampler in the next step
4. Handle the edge cases like when the draft model generates beyond `max_pos_embeddings`
5. Do E2E correctness and performance tests

I don't think I can do all these steps, so it'd be nice if we can get some helps from the community.

File: vllm/v1/spec_decode/eagle.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py]: Added TODO. I'd like to defer this to a future PR because 1) the sampling code in the PR needs some refactoring anyways, and 2) handling RNG in spec decoding is quite tricky, so I need more time to think more about it.

File: vllm/v1/spec_decode/eagle.py (hunk 6 above)
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py]: Added TODO. I'd like to defer this to a future PR because 1) the sampling code in the PR needs some refactoring anyways, and 2) handling RNG in spec decoding is quite tricky, so I need more time to think more about it.

**Comment by LiuXiaoxuanPKU** [comment]
Sounds good. I am happy to take 2 since I already start thinking about it. Let me create issues for 1,3,4,5,6 and move some of them to onboarding tasks.

**Comment by WoosukKwon** [comment]
@LiuXiaoxuanPKU Can you please take 1 as well?

**Review by sroy745 — COMMENTED** [review]
LGTM. One small comment.

**Review by LiuXiaoxuanPKU — APPROVED** [review]
LGTM.

File: vllm/v1/worker/gpu_model_runner.py [hunk 9]
```diff
@@ -1143,10 +1142,75 @@ def execute_model(
...
+        elif self.speculative_config.method == "eagle":
+            assert isinstance(self.drafter, EagleProposer)
+            # TODO(woosuk): Refactor the loop.
+            next_token_ids: list[int] = []
+            for i, token_ids in enumerate(valid_sampled_token_ids):
+                if token_ids:
+                    # Common case.
+                    next_token_id = token_ids[-1]
+                else:
+                    # Partial prefill (rare case).
```
[the remark this answers is not in the document]
Comment (LiuXiaoxuanPKU) [reply vllm/v1/worker/gpu_model_runner.py:1160]: this is for chunked context, eagle needs the next_token_id as the input. In the decoding phase, next token id is the last generated token id. In the chunked prefill case, it's the next token id of the current chunk.

File: vllm/v1/spec_decode/eagle.py [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+            input_ids = draft_token_ids_list[-1]
+            positions += 1
+            attn_metadata.max_seq_len += 1
+            attn_metadata.seq_lens += 1
+            # Compute the slot mapping.
+            block_numbers = positions // self.block_size
+            block_ids = block_table.gather(dim=1,
+                                           index=block_numbers.view(-1, 1))
+            block_ids = block_ids.view(-1)
+            attn_metadata.slot_mapping = (block_ids * self.block_size +
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py:113]: Unfortunately, we can't because if `slot mapping` might go beyond the block boundary.

File: vllm/config.py [hunk 11]
```diff
@@ -2153,9 +2153,9 @@ def __post_init__(self):
 
                 # Replace hf_config for EAGLE draft_model
                 if self.method == "eagle":
-                    if self.enable_chunked_prefill:
-                        raise ValueError(
-                            "Chunked prefill and EAGLE are not compatible.")
+                    # if self.enable_chunked_prefill:
+                    #     raise ValueError(
+                    #         "Chunked prefill and EAGLE are not compatible.")
```
Comment (DarkLight1337) [inline vllm/config.py:2158]: Should we remove this commented-out code?

Comment (WoosukKwon) [reply vllm/config.py]: Nice catch! Fixed.

**Comment by v-lmn** [comment]
I found a bug use this code
image

because self._num_decodes = 0 and decode_metadata is None
image

**Comment by ekagra-ranjan** [comment]
@v-lmn can you share how you ran this code? My understanding is that this code is not finished and cannot be executed until the futher PRs adding missing functionality in eagle
